### ISSUE TITLE
Issues running tests in parallel with Xcode 10

### ISSUE THREAD

**Issue by NachoSoto** [Quick/Quick#801]
- [x] I have read [CONTRIBUTING](https://github.com/Quick/Quick/blob/master/CONTRIBUTING.md) and have done my best to follow them.

This has been a a requested feature for a while (#383), and it's finally here with `Xcode 10.0`!
However looks like `Quick` will need some changes to support it correctly.

### What did you do?

My tests pass in Xcode `9.4` and `10.0 beta 1`, but not when enabling the new parallel option:
<img width="412" alt="screen shot 2018-06-17 at 10 12 06 am" src="https://user-images.githubusercontent.com/685609/41510311-fd615dee-7216-11e8-8309-a9d44ddfd8a0.png">

The failure occurs whether it's randomized or not.

### What did you expect to happen?

Tests pass.

### What actually happened instead?

An exception is thrown:
> 'context' cannot be used inside 'it', 'context' may only be used inside 'context' or 'describe'.

### Environment

 - Quick: [37ed3f5e28acc62ab8e2d612ea49a78f44c7b79a](https://github.com/Quick/Quick/commit/37ed3f5e28acc62ab8e2d612ea49a78f44c7b79a)
 - Nimble: [69bb1612cf74ae072fed2675eebe7dab51da4d26](https://github.com/Quick/Nimble/commit/69bb1612cf74ae072fed2675eebe7dab51da4d26)
 - Xcode Version: `Version 10.0 beta 1 (10L176w)`
 - Swift Version: `Apple Swift version 4.2 (swiftlang-1000.0.16.7 clang-1000.10.25.3)`

 - Carthage: `0.29.0`

**Comment by marcelofabri** [comment]
I've seen this as well, but this doesn't happen on every run. Here's the full stacktrace:

```
2018-06-18 19:53:54.575820-0700 xctest[43236:1370252] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'An exception occurred when building Quick's example groups.
Some possible reasons this might happen include:

- An 'expect(...).to' expectation was evaluated outside of an 'it', 'context', or 'describe' block
- 'sharedExamples' was called twice with the same name
- 'itBehavesLike' was called with a name that is not registered as a shared example

Here's the original exception: 'NSInternalInconsistencyException', reason: ''describe' cannot be used inside 'it', 'describe' may only be used inside 'context' or 'describe'. ', userInfo: '(null)''
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000107a031e6 __exceptionPreprocess + 294
	1   libobjc.A.dylib                     0x0000000107098031 objc_exception_throw + 48
	2   CoreFoundation                      0x0000000107a78975 +[NSException raise:format:] + 197
	3   Quick                               0x000000012207d1ea __23+[QuickSpec initialize]_block_invoke + 474
	4   Quick                               0x000000012209055a $SIyB_Ieg_TR + 10
	5   Quick                               0x000000012209a58d $SIyB_Ieg_TRTA + 13
	6   Quick                               0x000000012209a3d3 $S5Quick5WorldC30performWithCurrentExampleGroup_7closureyAA0fG0C_yyXEtF + 371
	7   Quick                               0x000000012209a508 $S5Quick5WorldC30performWithCurrentExampleGroup_7closureyAA0fG0C_yyXEtFTo + 152
	8   Quick                               0x000000012207cfea +[QuickSpec initialize] + 250
	9   libobjc.A.dylib                     0x0000000107098b04 CALLING_SOME_+initialize_METHOD + 19
	10  libobjc.A.dylib                     0x0000000107098e9e _class_initialize + 276
	11  libobjc.A.dylib                     0x000000010709f824 lookUpImpOrForward + 226
	12  libobjc.A.dylib                     0x00000001070af414 _objc_msgSend_uncached + 68
	13  libobjc.A.dylib                     0x0000000107096bd2 _class_resolveMethod + 124
	14  libobjc.A.dylib                     0x000000010709f918 lookUpImpOrForward + 470
	15  libobjc.A.dylib                     0x000000010709f717 lookUpImpOrNil + 20
	16  libobjc.A.dylib                     0x0000000107096f34 class_respondsToSelector + 37
	17  Foundation                          0x0000000106909885 +[NSBundle bundleForClass:] + 72
	18  Foundation                          0x000000010695bbcc -[NSBundle classNamed:] + 85
	19  Quick                               0x000000012208d78a $S5Quick020testCaseClassForTestC8WithName33_350B9DFD3A734676FC740155217B5F24LLyyXlXpSgSSF + 298
	20  Quick                               0x000000012208d5d3 $S5Quick0A24SelectedTestSuiteBuilderC03forC12CaseWithNameACSgSS_tcfc + 67
	21  Quick                               0x000000012208d52f $S5Quick0A24SelectedTestSuiteBuilderC03forC12CaseWithNameACSgSS_tcfC + 79
	22  Quick                               0x000000012208df48 $S5Quick0A9TestSuiteC08selectedbC003forB12CaseWithNameACSgSS_tFZ + 120
	23  Quick                               0x000000012208e1ec $S5Quick0A9TestSuiteC08selectedbC003forB12CaseWithNameACSgSS_tFZTo + 76
	24  Quick                               0x000000012207e4a7 +[XCTestSuite(QuickTestSuiteBuilder) qck_hooked_testSuiteForTestCaseWithName:] + 87
	25  XCTest                              0x0000000106f38fc4 +[XCTestSuite testClassSuitesForTestIdentifiers:skippingTestIdentifiers:testExecutionOrdering:] + 479
	26  XCTest                              0x0000000106fa8fe4 __102-[XCTTestRunSession executeTestIdentifiers:skippingTestIdentifiers:completionHandler:completionQueue:]_block_invoke + 80
	27  XCTest                              0x0000000106fa8c86 __44-[XCTTestRunSession runTestsAndReturnError:]_block_invoke.78 + 68
	28  XCTest                              0x0000000106f4eb3a -[XCTestObservationCenter _observeTestExecutionForBlock:] + 594
	29  XCTest                              0x0000000106fa88d9 -[XCTTestRunSession runTestsAndReturnError:] + 623
	30  XCTest                              0x0000000106f1d000 -[XCTestDriver runTestsAndReturnError:] + 422
	31  XCTest                              0x0000000106f9b69f _XCTestMain + 1480
	32  xctest                              0x00000001065cdf97 main + 284
	33  libdyld.dylib                       0x000000010b51a955 start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by marcelofabri** [comment]
From Xcode 10 Beta 2 release notes:

* Tests using a third-party testing framework that integrates with XCTest may experience test failures when running tests in parallel due to a test case class’s +initialize method being called on a non-main thread. (40939090)

**Workaround**: Disable parallel test execution.

**Comment by marcelofabri** [comment]
Adding `dispatch_async(dispatch_get_main_queue(), ^{ /* ... */ });` to `+[QuickSpec initialize]` seems to fix the problem. Should we have this workaround in the project? 🤔

**Comment by marcelofabri** [comment]
This should work now according to the Xcode 10 Beta 3 release notes:

* Tests using a third-party testing framework that integrates with XCTest now work properly when run in parallel. (40939090)

**Comment by ikesyo** [comment]
It's interesting that we will need #716 to make it working from what I tested.

**Comment by ikesyo** [comment]
Should be fixed by 

- Xcode 10 beta 3
- #716
- #806

**Comment by stonko1994** [comment]
I'm currently observing a similar issue in Xcode `10.2.1`. As soon as I enable the `Execute in parallel on Simulator` option no `Quick` tests get executed anymore.

I'm using quick `v2.1.0` and nimble `v8.0.2` on iOS with Objective-C tests only.

_Not sure if I should create a new issue or not as it seems to be related to this one._

**Comment by mickyzinho** [comment]
@stonko1994 Have you solved that problem? Did you open a new GitHub issue? We're experiencing similar problems with XCode 11.1.

**Comment by stonko1994** [comment]
@mickyzinho No, I was not able to find a fix but also did not open a new Issue.

**Comment by bvirlet** [comment]
@stonko1994 @mickyzinho I was experiencing this problem. As of Xcode 12.5 it seems that Quick tests are detected and run. (other issues though)
